src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs

```diff
@@ -587,7 +587,9 @@ private SubpatternElementSyntax ParseSubpatternElement()
         /// </summary>
         private bool IsPossibleSubpatternElement()
         {
-            return this.IsPossibleExpression() || this.CurrentToken.Kind == SyntaxKind.OpenBraceToken;
+            var tk = this.CurrentToken.Kind;
+            return this.IsPossibleExpression() && !(SyntaxFacts.IsBinaryExpression(tk) || SyntaxFacts.IsAssignmentExpressionOperatorToken(tk)) ||
+                this.CurrentToken.Kind == SyntaxKind.OpenBraceToken;
         }
```

Comment:
Out of curiosity, what was the loop's repeating stack trace?

Reply:
not to be "that guy".  But i think these sorts of complex expressions are hard to reason about (esp. in hte parser where there are so many cases to consider).  I think it would be much more desirable (but more verbose) to have something like:

```c#
if (!this.IsPossibleExpression()) {
   // not a possible subpattern element because...
   return false;
}

if (otherChecks) {
}
```

I've read the actual code above there a few times now and i cannot actually tell what the cases are that it's considering.  The combo of an ```&&```, ```||``` a ```!``` and a ```==``` actually overloaded me (and i just spent a long time on an entire IDE feature related to precedence in expressions :))

Reply:
For example, i keep on being confused if ```this.CurrentToken.Kind == SyntaxKind.OpenBraceToken``` means the entire expression is "true", no matter what anything else is.  Or if the expression could be false even in that case, if you didn't hav ea possible expression.

I believe it's the former.  In which case, why not do the simple "this.CurrentToken.Kind == SyntaxKind.OpenBraceToken" check first, and return true, if it succeed? (etc. etc.)

Reply:
@CyrusNajmabadi An expression *is* a possible subpattern element. Also, a subpattern can begin with an open curly brace. A subpattern is an optional identifier-colon, followed by a pattern.

@jcouv The infinite loop in the parser was in the following code, which adds more and more errors without consuming any tokens when the next token is `=`:

``` c#
                    while (true)
                    {
                        if (this.CurrentToken.Kind == SyntaxKind.CloseParenToken ||
                            this.CurrentToken.Kind == SyntaxKind.CloseBraceToken ||
                            this.CurrentToken.Kind == SyntaxKind.SemicolonToken)
                        {
                            break;
                        }
                        else if (this.CurrentToken.Kind == SyntaxKind.CommaToken || this.IsPossibleSubpatternElement())
                        {
                            list.AddSeparator(this.EatToken(SyntaxKind.CommaToken));
                            list.Add(this.ParseSubpatternElement());
                            continue;
                        }
                        else if (this.SkipBadPatternListTokens(ref openToken, list, SyntaxKind.CommaToken, closeKind) == PostSkipAction.Abort)
                        {
                            break;
                        }
                    }
```

Reply:
> @CyrusNajmabadi An expression is a possible subpattern element. Also, a subpattern can begin with an open curly brace. A subpattern is an optional identifier-colon, followed by a pattern.

Can we make that clearer in hte code but not having a large, complex, expression that attempts to check for all of those cases? :)

It would be particularly helpful to have an explanation of why you need the !(IsBinary || IsAssignment).  With separate cases, we can just put a nice example with a comment with each case demonstrating why it's checking what it's checking.

Reply:
@gafter Consider replacing the `while (true)` with the pattern to check that progress is made. Look for `IsMakingProgress()` helper.

Reply:
I don't do that check first because I believe it to be the least likely of the alternatives.

---
In reply to: [177844346](https://github.com/dotnet/roslyn/pull/25783#discussion_r177844346) [](ancestors = 177844346)

Reply:
The exclusions are due to the fact that `IsPossibleExpression` lies for those cases.

---
In reply to: [177902124](https://github.com/dotnet/roslyn/pull/25783#discussion_r177902124) [](ancestors = 177902124)